gst/mqtt/mqttsrc.c

```diff
@@ -485,6 +489,67 @@ gst_mqtt_src_get_caps (GstBaseSrc * basesrc, GstCaps * filter)
   return caps;
 }
 
+/**
+ * @brief Do negotiation procedure again if it needed
+ */
+static gboolean
+gst_mqtt_src_renegotiate (GstBaseSrc * basesrc)
+{
+  GstMqttSrc *self = GST_MQTT_SRC (basesrc);
+  GstCaps *caps = NULL;
+  GstCaps *peercaps = NULL;
+  GstCaps *thiscaps;
+  gboolean result = FALSE;
+  GstCaps *fixed_caps = NULL;
+
+  if (self->caps == NULL || gst_caps_is_any (self->caps))
+    goto no_nego_needed;
+
+  thiscaps = gst_pad_query_caps (GST_BASE_SRC_PAD (basesrc), NULL);
+  if (thiscaps && gst_caps_is_equal (self->caps, thiscaps)) {
+    gst_caps_unref (thiscaps);
+    goto no_nego_needed;
+  }
+
+  peercaps = gst_pad_peer_query_caps (GST_BASE_SRC_PAD (basesrc), self->caps);
+  if (peercaps) {
+    caps = gst_caps_ref (peercaps);
+    if (peercaps != self->caps)
+      gst_caps_unref (peercaps);
+  } else {
+    caps = gst_caps_ref (self->caps);
+  }
+
+  if (caps && !gst_caps_is_empty (caps)) {
+    if (gst_caps_is_any (caps)) {
+      result = TRUE;
+    } else {
+      fixed_caps = gst_caps_fixate (caps);
+
+      if (fixed_caps && gst_caps_is_fixed (fixed_caps)) {
+        result = gst_base_src_set_caps (basesrc, fixed_caps);
+        if (peercaps == self->caps)
+          gst_caps_unref (fixed_caps);
+      }
+    }
+    gst_caps_unref (caps);
+  } else {
+    result = FALSE;
```

Comment:
Q. no need to unref caps when `gst_caps_is_empty()`?

Reply:
It is required. Thanks.